packages/integrations/markdoc/src/index.ts

```diff
@@ -0,0 +1,127 @@
+import type { AstroIntegration, AstroConfig, ContentEntryType, HookParameters } from 'astro';
+import { InlineConfig } from 'vite';
+import type { Config } from '@markdoc/markdoc';
+import Markdoc from '@markdoc/markdoc';
+import {
+	prependForwardSlash,
+	getAstroConfigPath,
+	MarkdocError,
+	parseFrontmatter,
+} from './utils.js';
+import { fileURLToPath } from 'node:url';
+import fs from 'node:fs';
+
+type IntegrationWithPrivateHooks = {
+	name: string;
+	hooks: Omit<AstroIntegration['hooks'], 'astro:config:setup'> & {
+		'astro:config:setup': (
+			params: HookParameters<'astro:config:setup'> & {
+				// `contentEntryType` is not a public API
+				// Add type defs here
+				addContentEntryType: (contentEntryType: ContentEntryType) => void;
+			}
+		) => void | Promise<void>;
+	};
+};
+
+export default function markdoc(markdocConfig: Config = {}): IntegrationWithPrivateHooks {
+	return {
+		name: '@astrojs/markdoc',
+		hooks: {
+			'astro:config:setup': async ({ updateConfig, config, addContentEntryType }) => {
+				function getEntryInfo({ fileUrl, contents }: { fileUrl: URL; contents: string }) {
+					const parsed = parseFrontmatter(contents, fileURLToPath(fileUrl));
+					return {
+						data: parsed.data,
+						body: parsed.content,
+						slug: parsed.data.slug,
+						rawData: parsed.matter,
+					};
+				}
+				addContentEntryType({
+					extensions: ['.mdoc'],
+					getEntryInfo,
+					contentModuleTypes: await fs.promises.readFile(
+						new URL('../template/content-module-types.d.ts', import.meta.url),
+						'utf-8'
+					),
+				});
+
+				const viteConfig: InlineConfig = {
+					plugins: [
+						{
+							name: '@astrojs/markdoc',
+							async transform(code, id) {
+								if (!id.endsWith('.mdoc')) return;
+
+								validateRenderProperties(markdocConfig, config);
+								const body = getEntryInfo({
+									// Can't use `pathToFileUrl` - Vite IDs are not plain file paths
+									fileUrl: new URL(prependForwardSlash(id), 'file://'),
+									contents: code,
+								}).body;
+								const ast = Markdoc.parse(body);
+								const content = Markdoc.transform(ast, markdocConfig);
+
+								return `import { jsx as h } from 'astro/jsx-runtime';\nimport { Renderer } from '@astrojs/markdoc/components';\nconst transformedContent = ${JSON.stringify(
```

Comment:
Are we using the `jsx-runtime` here for any reason other than it having a convenient signature? I would expect this to use our JS API for components.